text/3064-bitfield.md

```diff
@@ -0,0 +1,384 @@
+struct are possible unless all accesses are immutable. Therefore, unlike in C,
+the compiler can implement access to bit-fields with any kind of load/write
+instruction, even if such an instruction overlaps the memory locations of other
+fields in the struct.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Bit-fields cannot be assigned without overflow checks.
+- The annotation feels somewhat far away from the type itself.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Alternative: Using `u32: N` to specify the width of a bit-field. I've chosen
+  to use an annotation instead because it seems more likely to be accepted.
+
+- Alternative: Using a dedicated type instead of an annotation:
+
+    ```rust
+    #[repr(C)]
+    struct X {
+        a: BitField<u32, 6>,
+    }
+    ```
+
+    where `BitField` would be some kind of lang item that ensures that
+
+    - the address of a field with such a type cannot be taken
+    - the type and width parameters are correct.
+
+    This type would then have the following API:
+
+    ```rust
+    impl<T, const N: usize> BitField<T, N> {
+        fn new(t: T) -> Self;
+        fn wrapping_new(t: T) -> Self;
+        fn get(self) -> T;
+        fn set(&mut self, t: T);
+        fn wrapping_add(self, t: T) -> Self;
+        ...
+    }
+
+    impl<T, const N: usize> AddAssign<T> for BitField<T, N> {
+    }
+    ```
+
+    The layout of `BitField<T, N>` would be the layout of `T` when used outside
+    a struct.
+
+    This could then be used like so:
+
+    ```rust
+    fn f(x: X) {
+        x.a.set(1);
+        x.a += 1;
+        x.a = x.a.wrapping_add(12);
+        println!("{}", x.a.get());
+    }
+    ```
+
+    There is some magic happening here:
+
+    - `x.a += 1` cannot be written as `AddAssign::add(&mut x.a, 1)` because the
+      address of `x.a` cannot be taken. The compiler would figure it out when
+      the lhs is a struct field.
+    - When the assignment happens, the compiler writes the bits on the rhs to
+      the correct positions in the struct.
+
+    This is a more structured solution but would need to be fleshed out.
+
+# Prior art
+[prior-art]: #prior-art
+
+- C
+- The author of Zig [proposes] a different syntax that basically boils down to
+  arbitrarily sized integer types:
+
+  ```rust
+  struct X {
+      x: i9,
+  }
+  ```
+
+  But this proposal is flawed because, in C, both the width and the underlying
+  type influence the layout. The Zig proposal throws the underlying type away.
+
+  ```c
+  struct X {
+      char a[3];
+      int b:9;
+  };
+
+  struct Y {
+      char a[3];
+      long b:9;
+  };
+  ```
+
+  These structs have different layouts on `x86_64-unknown-linux-gnu`.
+
+  [proposes]: (https://andrewkelley.me/post/a-better-way-to-implement-bit-fields.html)
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- On Windows, Clang and GCC produce different layouts for packed structs:
```

Comment:
We should specify, explicitly, that the behavior on MSVC targets must match that of MSVC.

(We don't, today, need to provide a means of controlling the `ms_struct` flag. We may need to eventually, though.)

Reply:
@retep998 Please give your opinion on which language, if any, should be added to this RFC and which language should be included in the reference.